### PR TITLE
TFP: seed sensitivity diagnostic — is seed=0 a lucky init?

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The TFP champion (val_primary/field_mse = 0.002383, PR #3025, seed=0) may be a lucky initialization. Seeds 42, 123, and 456 all diverged to infinity (issue #3168), raising the question of whether the baseline is reproducible or an outlier. This diagnostic maps the failure mechanism before we build further on this config.

Four cheap 20-epoch trials test whether the divergence is due to the seed itself, gradient norm explosion at init, missing warmup, or pressure-target scaling sensitivity.

## Instructions

All trials use the TFP champion config as the mandatory base:
Lion lr=1.25e-4, T_max=10, gc=0.5, WD=1e-2, EMA=0.999, 3L/192d, Fourier+physics, 360-min budget.

**For each trial, report:**
1. Finite or NaN/inf (and at which epoch divergence occurs if it diverges)
2. val_primary/field_mse at epoch 20 (or last finite epoch)
3. Grad norm behavior — stable, growing, or spiking

Use `--wandb_group tfp-seed-sensitivity` for all runs.

---

**Trial 1 — Control: verify seed=0 still converges (20 epochs)**
```bash
cd target/icml2026 && python train.py \
  --dataset tandemfoil_paper \
  --optimizer lion --lr 1.25e-4 \
  --cosine-t-max 10 --grad-clip 0.5 \
  --weight-decay 1e-2 --ema-decay 0.999 \
  --enable-fourier --model-layers 3 \
  --model-hidden-dim 192 --model-heads 3 \
  --enable-te-coord-frame --enable-cp-panel \
  --enable-cp-panel-tandem-only \
  --asinh-pressure --residual-prediction \
  --enable-pressure-prior-addition \
  --epochs 20 --seed 0 \
  --wandb_group tfp-seed-sensitivity \
  --wandb_run_name tfp-seed0-control
```

---

**Trial 2 — Tighter gradient clipping: seed=42, gc=0.3**
```bash
cd target/icml2026 && python train.py \
  --dataset tandemfoil_paper \
  --optimizer lion --lr 1.25e-4 \
  --cosine-t-max 10 --grad-clip 0.3 \
  --weight-decay 1e-2 --ema-decay 0.999 \
  --enable-fourier --model-layers 3 \
  --model-hidden-dim 192 --model-heads 3 \
  --enable-te-coord-frame --enable-cp-panel \
  --enable-cp-panel-tandem-only \
  --asinh-pressure --residual-prediction \
  --enable-pressure-prior-addition \
  --epochs 20 --seed 42 \
  --wandb_group tfp-seed-sensitivity \
  --wandb_run_name tfp-seed42-gc0.3
```

---

**Trial 3 — Warmup: seed=42, --warmup-epochs 3**
```bash
cd target/icml2026 && python train.py \
  --dataset tandemfoil_paper \
  --optimizer lion --lr 1.25e-4 \
  --cosine-t-max 10 --grad-clip 0.5 \
  --weight-decay 1e-2 --ema-decay 0.999 \
  --enable-fourier --model-layers 3 \
  --model-hidden-dim 192 --model-heads 3 \
  --enable-te-coord-frame --enable-cp-panel \
  --enable-cp-panel-tandem-only \
  --asinh-pressure --residual-prediction \
  --enable-pressure-prior-addition \
  --epochs 20 --seed 42 \
  --warmup-epochs 3 \
  --wandb_group tfp-seed-sensitivity \
  --wandb_run_name tfp-seed42-warmup3
```

---

**Trial 4 — Pressure transform: seed=42, --asinh-scale 0.5**
```bash
cd target/icml2026 && python train.py \
  --dataset tandemfoil_paper \
  --optimizer lion --lr 1.25e-4 \
  --cosine-t-max 10 --grad-clip 0.5 \
  --weight-decay 1e-2 --ema-decay 0.999 \
  --enable-fourier --model-layers 3 \
  --model-hidden-dim 192 --model-heads 3 \
  --enable-te-coord-frame --enable-cp-panel \
  --enable-cp-panel-tandem-only \
  --asinh-pressure --asinh-scale 0.5 \
  --residual-prediction \
  --enable-pressure-prior-addition \
  --epochs 20 --seed 42 \
  --wandb_group tfp-seed-sensitivity \
  --wandb_run_name tfp-seed42-asinh0.5
```

---

**Goal:** Understand the divergence mechanism, not beat the baseline. If Trial 1 itself diverges, that is the most important finding — the champion is not reproducible even on seed=0.

If any seed=42 trial stays finite at epoch 20, note whether val_primary/field_mse is within 2x of 0.002383 — that would suggest the config is recoverable with a stabilization fix.

## Baseline

| Metric | Value |
|---|---|
| val_primary/field_mse | **0.002383** |
| Epoch | 443 |
| PR | #3025 (haku/tfp-lion-champion-config) |
| W&B run | d1xh0o1p |
| Seed | 0 |

Known: divergence at ep462 due to T_max=10 cosine instability; EMA=0.999 preserved ep443 best.

Reproduce command:
```bash
cd target/icml2026 && python train.py \
  --dataset tandemfoil_paper \
  --optimizer lion --lr 1.25e-4 \
  --cosine-t-max 10 --grad-clip 0.5 \
  --weight-decay 1e-2 --enable-fourier \
  --model-layers 3 --model-hidden-dim 192 --model-heads 3 \
  --enable-te-coord-frame --enable-cp-panel \
  --enable-cp-panel-tandem-only \
  --asinh-pressure --residual-prediction \
  --enable-pressure-prior-addition \
  --epochs 999 --ema-decay 0.999
```

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT jin:

## Results

### Summary

All 4 trials show **val_primary/field_mse = Infinity** at all 20 epochs — the pressure channel diverges to Infinity across all seeds and all stabilization variants. This is **not seed-specific** — the divergence is systematic and present even in the seed=0 control that previously produced the champion result.

The Ux and Uy velocity channels improve normally throughout training; only the pressure channel (p) produces Infinity values. The `sinh` inverse transform used at evaluation time amplifies small errors in the asinh-compressed representation into astronomically large values, producing Infinity or ~1e+26–1e+32 in the raw pressure MSE.

### Detailed Results

| Trial | Seed | Variant | Ep20 val field_mse | Ep20 val surface_p | Ep20 val Ux | Ep20 val Uy | Ep20 loss | Ep20 grad_norm | Test field_mse |
|---|---|---|---|---|---|---|---|---|---|
| 1 (control) | 0 | gc=0.5 (baseline) | **Inf** | Inf | 0.004182 | 0.014554 | 0.1065 | 20.0 | **Inf** |
| 2 | 42 | gc=0.3 (tighter clip) | **Inf** | Inf | 0.004067 | 0.014116 | 0.0959 | 21.0 | **Inf** |
| 3 | 42 | warmup=3ep, gc=0.5 | **Inf** | 1.43e+29 | 0.004376 | 0.016050 | 0.1056 | 27.5 | **Inf** |
| 4 | 42 | asinh-scale=0.5, gc=0.5 | **Inf** | Inf | 0.004214 | 0.012653 | 0.0889 | 17.6 | **Inf** |

### Per-trial observations

1. **Trial 1 (seed=0 control):** The seed=0 champion config **does not converge** in 20 epochs on the current codebase. Pressure is Infinity from epoch 1 through 20. Grad norms stable at 20-27, 358-360/360 steps clipped. The original champion result at 0.002383 (ep443) was apparently achieved on an older version of the code.

2. **Trial 2 (seed=42, gc=0.3):** Tighter gradient clipping does not help. Pressure remains Infinity. One flicker at ep12 (surface_mse_p = 2.66e+32 — finite but enormous). Test surface_mse_p = 2.27e+26.

3. **Trial 3 (seed=42, warmup=3):** Linear warmup over 3 epochs shows occasional finite-but-huge pressure values (ep15: 5.09e+30, ep18: 8.14e+32, ep19: 2.20e+28, ep20: 1.43e+29). The warmup slightly delays pressure divergence but does not prevent it. Grad norms consistently higher (~27-32 vs ~20-21 for others).

4. **Trial 4 (seed=42, asinh-scale=0.5):** Lower asinh scale shows lowest training loss (0.089) and lowest grad norms (17.6). Occasional finite pressure values (ep12: 6.39e+30, ep14: 4.52e+26) but still Infinity overall. The halved scale slightly reduces pressure divergence magnitude but doesn't solve the root cause.

### Key findings

1. **The divergence is NOT seed-specific.** Seed=0 (the champion seed) also shows Infinity pressure from the start. This rules out the hypothesis that seed=0 was a "lucky init."

2. **The divergence is systematic.** None of the stabilization variants (tighter clipping, warmup, different asinh scale) prevent it. The pressure channel is fundamentally broken in the current codebase.

3. **The root cause is likely a code regression.** The champion result (0.002383 at ep443, W&B run d1xh0o1p) was produced on an earlier version of train.py. Something changed in the evaluation metric computation or the sinh inverse transform that now produces Infinity for the pressure channel.

4. **Velocity channels are healthy.** Ux/Uy losses decrease normally across all trials, suggesting the model architecture and training loop work fine for non-pressure targets.

### Bug fix included in this PR

Fixed a `primary_metric_key` scoping bug where a local variable assignment on lines 1657-1660 shadowed the `primary_metric_key()` function, causing `UnboundLocalError`. This was blocking all TFP training runs. Also added the `--warmup-epochs` flag (LinearLR warmup chained with CosineAnnealingLR via SequentialLR).

### Commands used

```bash
# Trial 1: seed=0 control
CUDA_VISIBLE_DEVICES=0 python train.py --dataset tandemfoil_paper --optimizer lion --lr 1.25e-4 --cosine-t-max 10 --grad-clip 0.5 --weight-decay 1e-2 --ema-decay 0.999 --enable-fourier --model-layers 3 --model-hidden-dim 192 --model-heads 3 --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction --enable-pressure-prior-addition --epochs 20 --seed 0 --wandb-group tfp-seed-sensitivity --wandb-name "jin/tfp-seed0-control" --agent jin

# Trial 2: seed=42, gc=0.3
CUDA_VISIBLE_DEVICES=1 python train.py --dataset tandemfoil_paper --optimizer lion --lr 1.25e-4 --cosine-t-max 10 --grad-clip 0.3 --weight-decay 1e-2 --ema-decay 0.999 --enable-fourier --model-layers 3 --model-hidden-dim 192 --model-heads 3 --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction --enable-pressure-prior-addition --epochs 20 --seed 42 --wandb-group tfp-seed-sensitivity --wandb-name "jin/tfp-seed42-gc0.3" --agent jin

# Trial 3: seed=42, warmup 3 epochs
CUDA_VISIBLE_DEVICES=2 python train.py --dataset tandemfoil_paper --optimizer lion --lr 1.25e-4 --cosine-t-max 10 --grad-clip 0.5 --weight-decay 1e-2 --ema-decay 0.999 --enable-fourier --model-layers 3 --model-hidden-dim 192 --model-heads 3 --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only --asinh-pressure --residual-prediction --enable-pressure-prior-addition --epochs 20 --seed 42 --warmup-epochs 3 --wandb-group tfp-seed-sensitivity --wandb-name "jin/tfp-seed42-warmup3" --agent jin

# Trial 4: seed=42, asinh-scale=0.5
CUDA_VISIBLE_DEVICES=3 python train.py --dataset tandemfoil_paper --optimizer lion --lr 1.25e-4 --cosine-t-max 10 --grad-clip 0.5 --weight-decay 1e-2 --ema-decay 0.999 --enable-fourier --model-layers 3 --model-hidden-dim 192 --model-heads 3 --enable-te-coord-frame --enable-cp-panel --enable-cp-panel-tandem-only --asinh-pressure --asinh-scale 0.5 --residual-prediction --enable-pressure-prior-addition --epochs 20 --seed 42 --wandb-group tfp-seed-sensitivity --wandb-name "jin/tfp-seed42-asinh0.5" --agent jin
```

### Peak memory usage

~47.5 GB VRAM per GPU (consistent across all 4 trials)

### W&B run IDs

- Trial 1: `65q3qcbo`
- Trial 2: `6hv5zqio`
- Trial 3: `g3hvoya7`
- Trial 4: `e2arojym`

All runs in group `tfp-seed-sensitivity`.

### Suggested follow-ups

1. **Bisect the code regression.** Compare the current `train.py` evaluation metric computation against the version that produced d1xh0o1p. The champion's 0.002383 was real (logged in W&B) — something broke since then.
2. **Check the sinh inverse transform for numerical overflow.** The evaluation code applies `sinh(x) / asinh_scale` which can overflow to Infinity for even moderately large `x`. The bf16 training may produce asinh-space predictions that are finite but large enough to overflow when `sinh()` is applied.
3. **Try disabling asinh pressure entirely** as a control to isolate whether the asinh transform itself is the source of the evaluation Infinity.

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #3205 because ADVISOR: CRITICAL finding — seed=0 control no longer reproduces the champion. All 4 trials (seed=0 baseline, seed=42 with gc=0.3/warmup/asinh=0.5) produce Inf on pressure. This is a CODE REGRESSION, not seed sensitivity — the champion config worked on an older train.py. Velocity channels are healthy (Ux~0.004, Uy~0.014), confirming the pathology is isolated to sinh inverse pressure transform at eval time. Waiting for sanji's cp_panel bug fix (#3209) before any further TFP optimization. Thank you for the thorough diagnostic and the primary_metric_key bug fix, jin..
